### PR TITLE
Live Activity Support

### PR DESCRIPTION
**Pull request by fhboswell**
# Description
## One Line Summary
Adds Live Activity support for the SDK

### Motivation
Adding 2 sdk methods that allow a customer to register their temporary live activity token with an activity ID on OneSignal's backend

### Scope
Adding two SDK public methods and two corresponding requests

# Testing
## Unit testing
Added several tests to test the requests and the public SDK method

## Manual testing
The backend piece is not available yet so only testing with mock data.

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [x] Public API changes

# Checklist
## Overview
   - [ x] I have filled out all **REQUIRED** sections above
   - [x] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [x] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [x] I have included test coverage for these changes, or explained why they are not needed
   - [x] All automated tests pass, or I explained why that is not possible
   - [x] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [x] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [x] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

- - -
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152)

### CODE REVIEW

**Review by emawby — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152)* status: 0 of 6 files reviewed, 5 unresolved discussions (waiting on @fhboswell, @jennantilla, @jkasten2, and @nan-li)

___
*[`iOS_SDK/OneSignalSDK/OneSignalCore/Source/API/OSRequests.m` line 500 at r1](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGxWstF6lIDAnfqqML3:-NGxWstF6lIDAnfqqML4:bx12poi) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/e3a4435e95939a68a2a382f61a808c5c69a1d795/iOS_SDK/OneSignalSDK/OneSignalCore/Source/API/OSRequests.m#L500)):*
> ```objc
>                      appId:(NSString * _Nonnull)appId
>                 activityId:(NSString * _Nonnull)activityId
>                      token:(NSString * _Nullable)token {
> ```

Design question - Are we sure the token should be nullable? Is this so that folks can track a live activity with OneSignal even if they aren't updating it via push?
___
*[`iOS_SDK/OneSignalSDK/Source/OneSignal.h` line 298 at r1](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGxXFlc2LAgMW3xzI3e:-NGxXFlc2LAgMW3xzI3f:bmss72w) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/e3a4435e95939a68a2a382f61a808c5c69a1d795/iOS_SDK/OneSignalSDK/Source/OneSignal.h#L298)):*
> ```objc
> 
> #pragma mark Live Activity
> typedef void (^OSLiveActivitySuccessBlock)();
> ```

Design question - Is there any server response info we want to pass as a parameter to this? My gut reaction is that there isn't, but thought I would ask.
___
*[`iOS_SDK/OneSignalSDK/Source/OneSignal.m` line 671 at r1](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGxYLTW9R0Qq8SyJCiY:-NGxYLTW9R0Qq8SyJCiZ:b-rcdhc4) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/e3a4435e95939a68a2a382f61a808c5c69a1d795/iOS_SDK/OneSignalSDK/Source/OneSignal.m#L671)):*
> ```objc
> }
> + (void)enterLiveActivity:(NSString * _Nonnull)activityId withToken:(NSString * _Nullable)token withSuccess:(OSLiveActivitySuccessBlock _Nullable)successBlock withFailure:(OSLiveActivityFailureBlock _Nullable)failureBlock{
>     
> ```

We probably want to check privacy consent here (see sendTags for an example)
___
*[`iOS_SDK/OneSignalSDK/Source/OneSignal.m` line 672 at r1](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGxXeDMDC9KNX8ydsj_:-NGxXeDMDC9KNX8ydsja:b1vsyuo) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/e3a4435e95939a68a2a382f61a808c5c69a1d795/iOS_SDK/OneSignalSDK/Source/OneSignal.m#L672)):*
> ```objc
> + (void)enterLiveActivity:(NSString * _Nonnull)activityId withToken:(NSString * _Nullable)token withSuccess:(OSLiveActivitySuccessBlock _Nullable)successBlock withFailure:(OSLiveActivityFailureBlock _Nullable)failureBlock{
>     
>     [OneSignalClient.sharedClient executeRequest:[OSRequestLiveActivityEnter withUserId:self.currentSubscriptionState.userId appId:appId activityId:activityId token:token]
> ```

What happens if we don't have the userId yet? We should probably queue it up as a pending call to make once register user has finished successfully. See tags as an example.
___
*[`iOS_SDK/OneSignalSDK/Source/OneSignal.m` line 685 at r1](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGxYbAM5eI0EYAyaH-B:-NGxYbAM5eI0EYAyaH-C:b-c4rsi1) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/e3a4435e95939a68a2a382f61a808c5c69a1d795/iOS_SDK/OneSignalSDK/Source/OneSignal.m#L685)):*
> ```objc
> 
> + (void)exitLiveActivity:(NSString * _Nonnull)activityId withSuccess:(OSLiveActivitySuccessBlock _Nullable)successBlock withFailure:(OSLiveActivityFailureBlock _Nullable)failureBlock{
> ```

same comments for Exit as in Enter

File: iOS_SDK/OneSignalSDK/Source/OneSignal.h [hunk 1]
```diff
@@ -293,6 +293,17 @@ typedef void (^OSFailureBlock)(NSError* error);
 + (void)setLaunchURLsInApp:(BOOL)launchInApp;
 + (void)setProvidesNotificationSettingsView:(BOOL)providesView;
 
+
+#pragma mark Live Activity
+typedef void (^OSLiveActivitySuccessBlock)();
+typedef void (^OSLiveActivityFailureBlock)(NSError *error);
+
++ (void)enterLiveActivity:(NSString * _Nonnull)activityId withToken:(NSString * _Nullable)token;
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/Source/OneSignal.h:301]: `token` should be `_Nonnull`, as it is required.

File: iOS_SDK/OneSignalSDK/Source/OneSignal.h [hunk 2]
```diff
@@ -293,6 +293,17 @@ typedef void (^OSFailureBlock)(NSError* error);
...
++ (void)enterLiveActivity:(NSString * _Nonnull)activityId withToken:(NSString * _Nullable)token withSuccess:(OSLiveActivitySuccessBlock _Nullable)successBlock withFailure:(OSLiveActivityFailureBlock _Nullable)failureBlock;
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/Source/OneSignal.h:302]: same `_Nonnull` as noted above.

File: iOS_SDK/OneSignalSDK/UnitTests/RequestTests.m [hunk 3]
```diff
@@ -106,6 +110,7 @@ - (void)setUp {
 
     testAction = testBridgeEvent.userAction;
     testAction.firstClick = true;
+
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/UnitTests/RequestTests.m:113]: nit, remove extra added line.

File: iOS_SDK/OneSignalSDK/OneSignalCore/Source/API/OSRequests.m [hunk 4]
```diff
@@ -492,3 +492,30 @@ + (instancetype)trackUsageData:(NSString *)osUsageData appId:(NSString *)appId {
...
+    request.method = POST;
+    request.path = [NSString stringWithFormat:@"apps/%@/live_activities/%@/token", appId, activityId];
+    return request;
+}
+@end
+
+@implementation OSRequestLiveActivityExit
++ (instancetype)withUserId:(NSString * _Nonnull)userId
+                     appId:(NSString * _Nonnull)appId
+                activityId:(NSString * _Nonnull)activityId{
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/OneSignalCore/Source/API/OSRequests.m:515]: nit, should be a space after `activityId` and before `{`

**Review by jkasten2 — CHANGES_REQUESTED** [review]

**Review by fhboswell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152)* status: 0 of 6 files reviewed, 7 unresolved discussions (waiting on @emawby, @fhboswell, @jennantilla, @jkasten2, and @nan-li)

___
*[`iOS_SDK/OneSignalSDK/OneSignalCore/Source/API/OSRequests.m` line 500 at r1](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGxWstF6lIDAnfqqML3:-NGxapIS6xZE6K85Yvpv:b-xqiez0) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/e3a4435e95939a68a2a382f61a808c5c69a1d795/iOS_SDK/OneSignalSDK/OneSignalCore/Source/API/OSRequests.m#L500)):*
<details><summary><i>Previously, emawby (Elliot Mawby) wrote…</i></summary><blockquote>

Design question - Are we sure the token should be nullable? Is this so that folks can track a live activity with OneSignal even if they aren't updating it via push?
</blockquote></details>

It should not be nullable
___
*[`iOS_SDK/OneSignalSDK/Source/OneSignal.h` line 298 at r1](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGxXFlc2LAgMW3xzI3e:-NGxZd_21YScdSH0XOqh:bm1avd0) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/e3a4435e95939a68a2a382f61a808c5c69a1d795/iOS_SDK/OneSignalSDK/Source/OneSignal.h#L298)):*
<details><summary><i>Previously, emawby (Elliot Mawby) wrote…</i></summary><blockquote>

Design question - Is there any server response info we want to pass as a parameter to this? My gut reaction is that there isn't, but thought I would ask.
</blockquote></details>

Neither API endpoint has anything data in the response, see tech spec.
___
*[`iOS_SDK/OneSignalSDK/Source/OneSignal.h` line 301 at r1](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGwZ8rsf03GFXoHo7Sh-r1-301:-NGxaU1Z7MG0p_VCScg7:b-896fix) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/e3a4435e95939a68a2a382f61a808c5c69a1d795/iOS_SDK/OneSignalSDK/Source/OneSignal.h#L301)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

`token` should be `_Nonnull`, as it is required.
</blockquote></details>

Done.
___
*[`iOS_SDK/OneSignalSDK/Source/OneSignal.h` line 302 at r1](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGwZ8rsf03GFXoHo7Sh-r1-302:-NGxaUUg9n_Q7foqbvqV:b-896fix) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/e3a4435e95939a68a2a382f61a808c5c69a1d795/iOS_SDK/OneSignalSDK/Source/OneSignal.h#L302)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

same `_Nonnull` as noted above.
</blockquote></details>

Done.
___
*[`iOS_SDK/OneSignalSDK/Source/OneSignal.m` line 671 at r1](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGxYLTW9R0Qq8SyJCiY:-NGxcSnnE1OwxrpPmyYH:b-896fix) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/e3a4435e95939a68a2a382f61a808c5c69a1d795/iOS_SDK/OneSignalSDK/Source/OneSignal.m#L671)):*
<details><summary><i>Previously, emawby (Elliot Mawby) wrote…</i></summary><blockquote>

We probably want to check privacy consent here (see sendTags for an example)
</blockquote></details>

Done.

**Review by emawby — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152)* status: 0 of 6 files reviewed, 2 unresolved discussions (waiting on @jennantilla, @jkasten2, and @nan-li)

**Review by jkasten2 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 6 files at r1, 2 of 4 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152)* status: all files reviewed, 2 unresolved discussions (waiting on @fhboswell, @jennantilla, and @nan-li)

___
*[`iOS_SDK/OneSignalSDK/OneSignalCore/Source/API/OSRequests.m` line 500 at r3](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGy8cezAS5LOAMZqq0W:-NGy8cezAS5LOAMZqq0X:bgntf1r) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/63a1b9c8828216a2b54be7f1d9bb475aeffa2de5/iOS_SDK/OneSignalSDK/OneSignalCore/Source/API/OSRequests.m#L500)):*
> ```objc
>                      appId:(NSString * _Nonnull)appId
>                 activityId:(NSString * _Nonnull)activityId
>                      token:(NSString * _Nullable)token {
> ```

token here should be `_Nonnull`.
___
*[`iOS_SDK/OneSignalSDK/Source/OneSignal.m` line 722 at r3](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGy9Mc5A-0uTXes14u3:-NGy9Mc6E8N-ESZIrAIA:b-qpf3gu) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/63a1b9c8828216a2b54be7f1d9bb475aeffa2de5/iOS_SDK/OneSignalSDK/Source/OneSignal.m#L722)):*
> ```objc
> 
> + (void)executePendingLiveActivityUpdates{
>     if(pendingLiveActivityUpdates.count <= 0){
> ```

nit, missing spaces which doesn't match the rest of the formatting. Here as well as a few other places.

**Review by fhboswell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152)* status: 3 of 6 files reviewed, 1 unresolved discussion (waiting on @jennantilla, @jkasten2, and @nan-li)

___
*[`iOS_SDK/OneSignalSDK/OneSignalCore/Source/API/OSRequests.m` line 500 at r3](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-NGy8cezAS5LOAMZqq0W:-NH0BJNSERLo3Sb3W1tg:bb6un1j) ([raw file](https://github.com/onesignal/onesignal-ios-sdk/blob/63a1b9c8828216a2b54be7f1d9bb475aeffa2de5/iOS_SDK/OneSignalSDK/OneSignalCore/Source/API/OSRequests.m#L500)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

token here should be `_Nonnull`.
</blockquote></details>

Done

**Review by fhboswell — COMMENTED** [review]
> Dismissed @jkasten2 from a discussion.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152)* status: 3 of 6 files reviewed, all discussions resolved (waiting on @jennantilla, @jkasten2, and @nan-li)

**Review by jkasten2 — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/1152#-:-NH0Zlk1B74FlkFPtEJy:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @jennantilla and @nan-li)

**Pull request merged**